components/html-block/html-block.js

```diff
@@ -271,15 +283,23 @@ class HtmlBlock extends LitElement {
 			}
 		};
 
-		const template = this._findSlottedElement('TEMPLATE', slot);
+		if (this._contentObserver) this._contentObserver.disconnect();
 
-		if (this._templateObserver) this._templateObserver.disconnect();
+		const template = this._findSlottedElement('TEMPLATE', slot);
 		if (template) {
-			this._templateObserver = new MutationObserver(() => stampHTML(template));
-			this._templateObserver.observe(template.content, { attributes: true, childList: true, subtree: true });
+			this._contentObserver = new MutationObserver(() => stampHTML(template, true));
+			this._contentObserver.observe(template.content, { attributes: true, childList: true, subtree: true });
+			stampHTML(template, true);
+			return;
+		}
+
+		const container = this._findSlottedElement('DIV', slot);
+		if (container) {
+			this._contentObserver = new MutationObserver(() => stampHTML(container));
+			this._contentObserver.observe(container, { attributes: true, childList: true, subtree: true });
```

Comment:
Very naively: could we not just listen to the `slotchange` event instead of wiring `MutationObserver`s up to the things inside the slot, or are things weird because this doesn't actually use Lit's `render()`? Or is the `MutationObserver` so that we catch any changes the consumer might make to their content _after_ we've initially rendered it?

What might be nice is if consumers didn't even need to put the `<div>` in there (which is kind of a strange API), but instead could just put their HTML right inside. That would mean changing this code slightly though to wire up observers to all children and to call `stampHTML` once for each child -- but would have the advantage that if only part of their HTML changes, you only re-stamp that bit. 🤷

Reply:
So yes, I believe this was originally done via an observer because of the second reason you mention, since the `slotchange` event only fires if the slotted element itself changes, not its descendants. I don't believe it has anything to do with Lit's rendering in this case.

Yeah, I don't think it's unreasonable to do something like that and I agree that this API makes much less sense with a div vs. a template. I definitely worry a bit about the cost of stamping once for every immediate child of the html-block... My main concern is the MathJax renderer, which has to create a custom document and do a bunch of wire-up. I _think_ we can get around that by restructuring how the renderers work though. MathJax in particular supports typesetting lists of elements and I would hope it would be able to do so at least somewhat efficiently. 🤔

I do think I like the above idea better from an API point of view though, since fetching the original HTML just becomes a case of pulling the html-block's `innerHTML`.

Reply:
I guess my thought is that on initial render, we probably want to stamp all children at once, and then afterwards stamping via the observer could happen one element at a time. I guess my concern there would be if the consumer changed out literally everything (which the LMS for sure does). In that case, we run into a situation where we have little choice but to have MathJax typeset element-by-element, unless we get clever with some way of batching changes within the renderer.

Reply:
Though... The slotchange event handler could always batch updates, which would probably solve that problem. If someone changed out all of their HTML, it's the slotchange events that would be firing to indicate that change. 🤔

Reply:
I can undo any changes here in the PR easily; I'm going to wire something like this up and if we don't like it, I can back it out.

Reply:
>So yes, I believe this was originally done via an observer because of the second reason you mention, since the slotchange event only fires if the slotted element itself changes, not its descendants. I don't believe it has anything to do with Lit's rendering in this case.

I can confirm this is the case.

Reply:
Consumers might try specifying multiple sibling `div` elements into the slot, and if we aren't prepared for it we'd just ignore everything except the first one.  They may or may not notice depending on the breadth of their testing.  Not great.  

So, I'm also interested in enabling them to just put their content in without requiring a single wrapper `div`.  Having said that, we should be considerate of any perf differences, good or bad, since all the plugins would be running that many more times, albeit on smaller bits of HTML.  

Also, be careful of text nodes because they only have a subset of the element DOM API.  Not sure if this would break anything.
```html
<d2l-html-block>
   <div>stuff</div>
   some text
   <!-- some comment -->
   <div>more stuff</div>
</d2l-html-block>
```

Reply:
^ how many times would MathJax's typesetting be done in this case, and on what HTML would it be operating in each case?

Reply:
Good point RE: text nodes. I think each renderer would need to have some understanding of how to handle them. I think we could straight up ignore comment nodes (I imagine we wouldn't want to mutate them under any circumstances, since I assume whoever wrote the comment wants it to be as is), but text nodes could have math or @mentions in them.

Naively, I wonder about batching the slotchange event. i.e. The html-block stores some internal state mapping handlers with the element they're processing (I think that needs to be a thing lest one event fires twice and we end up rendering duplicates). We then batch those in some reasonable way and only run the renderers once. This would require changing the renderers to operate on a list of elements rather than one at a time though.

And yes, MathJax in particular would typeset once per element, and it would be operating only on that element and its subtree.

Reply:
I'm less convinced we need special handling for the observers, though we could do something similar and I guess we may as well. The main reason I'm less concerned is because the slotchange event will already cover the cases where elements are being added/removed, so the main issue there would be if some consumer added an attribute to every element within the html-block or something of that sort.

Reply:
For that initial render, you could also just take all the children and put them inside your own `<div>` before slamming them into the `<template>`'s `innerHTML` and then typeset that element.

Regardless, I think there's some combination of `slotchange` + mutation observers that should allow arbitrary children to be handled in a performant way... but likely requires some experimentation & testing. :)

Reply:
Yeah, it occurs to me that we're using the `slotchange` event to handle initial rendering anyways, so whatever works for the general case would also handle initial rendering. Naively, my thought is to batch via rAF - i.e. bundle up any events that process within a single frame and then render that batch at the end of the frame. I'm just starting to actually write code to test that all out though, so no word on how well it works quite yet. :D

Reply:
There's perhaps a broader problem with the idea of allowing rendering of only one or more children though. The MathJax renderer creates a new document and pulls in everything it's going to render. I don't know how feasible it is to adapt that to working on a list of potentially disparate elements scattered around the DOM while maintaining relative DOM order post-render.